### PR TITLE
Fix computation of view range (in blocks) sent to server

### LINKED ISSUE #4878: Block loading bug at small view range
How to reproduce:

1. go to an undergeround cave
2. log out
3. set view range to 20
4. log in

Underground blocks may not load on login.

### PR DESCRIPTION
**Pull request by Rogier-5**
Fixes #4878

### CODE REVIEW

**Comment by lhofhansl** [comment]
Can you elaborate? :)

With wanted_range = 20, you'd get 19/17 now and 20/16 before, both would end up being 1.

Does that fix the problem for you?

I think there is another problem with the server only sending with 1 map block send distance. I had thought the fix is to send at least with a map block radius of 2.

**Comment by juhdanad** [comment]
Yes, the bug is fixed.
@lhofhansl it is 19/16+1 now, which is 2.

**Comment by Rogier-5** [comment]
@est31 It does fix the issue I am seeing, and which I think is the bug @juhdanad reported.

With wanted_range_nodes = 20:
Previously: `wanted_range_blocks = 20/16 = 1`
With this fix: `wanted_range_blocks = 19/16 + 1 = 1 + 1 = 2`

Basically, I changed the formula from:
    `wanted_range_blocks = floor(wanted_range_nodes / 16.0)`
to
    `wanted_range_blocks = ceil(wanted_range_nodes / 16.0)`

Maybe I should add some more clarification...

File: src/client.cpp [hunk 1]
```diff
@@ -939,7 +939,7 @@ void writePlayerPos(LocalPlayer *myplayer, ClientMap *clientMap, NetworkPacket *
 	u32 keyPressed   = myplayer->keyPressed;
 	// scaled by 80, so that pi can fit into a u8
 	u8 fov           = clientMap->getCameraFov() * 80;
-	u8 wanted_range  = clientMap->getControl().wanted_range / MAP_BLOCKSIZE;
+	u8 wanted_range  = (clientMap->getControl().wanted_range - 1) / MAP_BLOCKSIZE + 1;
```
Comment (sfan5) [inline src/client.cpp]: don't we have a `ceil()` function for this?

**Comment by lhofhansl** [comment]
Oh. I see. Of course. 👍

**Comment by lhofhansl** [comment]
Now we can also change [this](https://github.com/minetest/minetest/blob/master/src/game.cpp#L4217), perhaps to 0.95.

**Comment by Rogier-5** [comment]
> Now we can also change this, perhaps to 0.95.

Won't 1.0 work ?

IIRC the fog formula reaches 100% fog before it reaches the max viewing range. So there should be no need to artificially reduce the viewing range.

**Comment by lhofhansl** [comment]
Yeah. 1.0 works not. I was wondering why that was still needed. You found the answer.
I just tried to simply remove it (i.e. 1.0) and it looks fine in all cases. Wanna just do that as part of this PR?

@paramat would like that too :)

**Comment by lhofhansl** [comment]
Cool. This all is coming together nicely. This is a great community.

**Comment by Rogier-5** [comment]
> Wanna just do that as part of this PR?

Well... I've already got the code. Just need to commit and push...

**Comment by lhofhansl** [comment]
@Rogier-5 I means removing the 0.9 fog range factor as part of this PR.

**Comment by lhofhansl** [comment]
Ah. You did already. Cool.

File: src/client.cpp (hunk 1 above)
Comment (Rogier-5) [inline src/client.cpp]: I just noticed that wanted_range is in fact already a float. I assumed it was integral...

Changed to ceil().

**Comment by paramat** [comment]
How about setting the min view range to 30?

**Comment by paramat** [comment]
This seems good.

File: src/game.cpp [hunk 2]
```diff
@@ -4214,7 +4214,7 @@ void Game::updateFrame(ProfilerGraph *graph, RunStats *stats,
 	if (draw_control->range_all) {
 		runData->fog_range = 100000 * BS;
 	} else {
-		runData->fog_range = 0.9 * draw_control->wanted_range * BS;
+		runData->fog_range = draw_control->wanted_range * BS;
```
Comment (paramat) [inline src/game.cpp]: I'm unsure about this, the 'wanted range' here is surely a diffrent local value in the client, the 'ceil' added above is only sent the server?

**Comment by paramat** [comment]
Looks good apart from the fog thing, see line comment.

**Comment by lhofhansl** [comment]
Tested this. I have not been able to produce any visible edges at any view range.

We discussed this on #4811, there I said I can see edges when this removed. Turns out that was due to me sending inadvertently floor of view-range / MAP_BLOCKSIZE. Due to that we'd retrieve sightly less on terms of range. I remember I was wondering back then; now I have an explanation. :)
Pretty certain this is fine.

File: src/game.cpp (hunk 2 above)
Comment (paramat) [inline src/game.cpp]: There's a high chance i'm wrong though.

**Comment by paramat** [comment]
If i'm wrong about the fog change then +1.

**Comment by paramat** [comment]
Will test.

**Comment by paramat** [comment]
Works :+1:

**Comment by lhofhansl** [comment]
We now have this and #4884, and some of the logic is duplicated.
Let's get this one in, and then do #4884 on top. (that'd be my personal preference at least)

**Comment by est31** [comment]
:+1:

**Pull request merged**